dali/operators/reader/loader/tar_utils_test.cc

```diff
@@ -0,0 +1,215 @@
+  TAR* handle;
+  ASSERT_EQ(tar_open(&handle, GetParam().filepath.c_str(), NULL, O_RDONLY, 0, TAR_GNU), 0);
+
+  do {
+    th_read(handle);
+    vector<uint8_t> libtar_contents;
+    uint64 count = 0;
+    uint64 filesize = th_get_size(handle);
+
+    do {
+      libtar_contents.resize(libtar_contents.size() + T_BLOCKSIZE);
+      tar_block_read(handle, libtar_contents.data() + libtar_contents.size() - T_BLOCKSIZE);
+    } while ((count += T_BLOCKSIZE) < filesize);
+    libtar_contents.resize(filesize);
+
+    ASSERT_EQ(archive.Read(), libtar_contents);
+  } while (archive.NextFile());
+  ASSERT_EQ(th_read(handle), 1);
+  ASSERT_EQ(tar_close(handle), 0);
+}
+
+auto SimpleTarTestsValues() {
+  vector<SimpleTarTestsData> values;
+
+  SimpleTarTestsData filepaths[] = {
+      {dali::filesystem::join_path(std::getenv("DALI_EXTRA_PATH"),
+                                   "db/webdataset/MNIST/devel-0.tar"),
+       false,
+       false,
+       2000,
+       3000,
+       {".cls", ".jpg"}},
+      {dali::filesystem::join_path(std::getenv("DALI_EXTRA_PATH"),
+                                   "db/webdataset/sample-tar/empty.tar"),
+       false,
+       false,
+       0,
+       0,
+       {}},
+      {dali::filesystem::join_path(std::getenv("DALI_EXTRA_PATH"),
+                                   "db/webdataset/sample-tar/v7.tar"),
+       false,
+       false,
+       0,
+       1000,
+       {""}},
+      {dali::filesystem::join_path(std::getenv("DALI_EXTRA_PATH"),
+                                   "db/webdataset/sample-tar/oldgnu.tar"),
+       false,
+       false,
+       0,
+       1000,
+       {""}}};
+
+  for (auto& filepath : filepaths) {
+    for (int read_ahead = 0; read_ahead <= 1; read_ahead++) {
+      for (int use_mmap = 0; use_mmap <= 1; use_mmap++) {
+        filepath.read_ahead = read_ahead;
+        filepath.use_mmap = use_mmap;
+        values.push_back(filepath);
+      }
+    }
+  }
+  return testing::ValuesIn(values.begin(), values.end());
+}
+
+INSTANTIATE_TEST_CASE_P(LibTarUtilsTestParametrized, SimpleTarTests, SimpleTarTestsValues());
+
+
+constexpr int kMultithreadedSamples = 3;
+
+class MultiTarTests : public ::testing::TestWithParam<bool> {
+ protected:
+  std::unique_ptr<TarArchive> archives[kMultithreadedSamples];
+  const std::pair<int, int> ranges[kMultithreadedSamples] = {
+    {2000, 3000},
+    {0, 1000},
+    {1000, 2000}
+  };
+  void SetUp() override {
+    std::string filepath_prefix(
+        dali::filesystem::join_path(std::getenv("DALI_EXTRA_PATH"), "db/webdataset/MNIST/devel-"));
+
+    std::string filepaths[kMultithreadedSamples] = {
+        filepath_prefix + "0.tar", filepath_prefix + "1.tar", filepath_prefix + "2.tar"};
+
+    for (int idx = 0; idx < kMultithreadedSamples; idx++) {
+      archives[idx] = std::make_unique<TarArchive>(FileStream::Open(filepaths[idx], false, false));
+    }
+  }
+};
+
+TEST_P(MultiTarTests, Index) {
+  std::future<void> threads[kMultithreadedSamples];
```

Comment:
Why call them threads instead of, say, tasks? I don't think std::async guarantees them to be in separate threads.

Reply:
I think they can use threads from a thread pool and there can be fewer of them than there are tasks.